File: algos/PY/mst.py
class UnionFind:
    def __init__(self, n):
        self.parent = list(range(n))
        self.rank = [0] * n

    def find(self, u):
        if self.parent[u] != u:
            self.parent[u] = self.find(self.parent[u])
        return self.parent[u]

    def union(self, u, v):
        root_u = self.find(u)
        root_v = self.find(v)

        if root_u != root_v:
            if self.rank[root_u] > self.rank[root_v]:
                self.parent[root_v] = root_u
            elif self.rank[root_u] < self.rank[root_v]:
                self.parent[root_u] = root_v
            else:
                self.parent[root_v] = root_u
                self.rank[root_u] += 1
    
class Kruskal:
    def __init__(self, v:int):
        self.v_cnt = v
        self.edges = []

    def add_edge(self, u, v, w):
        self.edges.append((w, u, v))

    def get_mst(self):
        uf = UnionFind(self.v_cnt)
        self.edges.sort()
        mst_weight = 0
        mst_edges = []
        for w, u, v in self.edges:
            if uf.find(u) != uf.find(v):
                uf.union(u, v)
                mst_weight += w
                mst_edges.append((u, v, w))
        return mst_weight, mst_edges
    
class Prim:
    def __init__(self, v:int):
        self.v_cnt = v
        self.graph = [[] for _ in range(v)]

    def add_edge(self, u, v, w):
        self.graph[u].append((w, v))
        self.graph[v].append((w, u))

    def get_mst(self):
        import heapq
        mst_weight = 0
        mst_edges = []
        visited = [False] * self.v_cnt
        min_heap = []
        visited[0] = True
        for w, v in self.graph[0]:
            heapq.heappush(min_heap, (w, v))

        while min_heap:
            w, v = heapq.heappop(min_heap)
            if not visited[v]:
                visited[v] = True
                mst_weight += w
                mst_edges.append((0, v, w))
                for next_w, next_v in self.graph[v]:
                    if not visited[next_v]:
                        heapq.heappush(min_heap, (next_w, next_v))
        return mst_weight, mst_edges

def mst_algo(V, E):
    floor_log_V = V.bit_length() - 1
    if E > (V * floor_log_V): # dense
        return Prim(V)
    else: # sparse
        return Kruskal(V)

File: algos/GO/mst.go
package algos

import (
	"math/bits"
	"sort"
)

type Edge[V number, W number] struct {
	From   V
	To     V
	Weight W
}

type MST[T number, W number] interface {
	// AddEdge adds an edge to the MST algorithm.
	AddEdge(from, to T, weight W)

	// GetMST returns the edges in the minimum spanning tree.
	GetMST() []Edge[T, W]
}

type Kruskal[V number, W number] struct {
	edges []Edge[V, W]
	uf    *UnionFind[V]
}
func (k *Kruskal[V, W]) AddEdge(from, to V, weight W) {
	k.edges = append(k.edges, Edge[V, W]{From: from, To: to, Weight: weight})
}
func (k *Kruskal[V, W]) GetMST() []Edge[V, W] {
	if len(k.edges) == 0 {
		return nil	}

	// Sort edges by weight
	sort.Slice(k.edges, func(i, j int) bool {
		return k.edges[i].Weight < k.edges[j].Weight
	})
	
	mst := []Edge[V, W]{}
	for _, edge := range k.edges {
		if k.uf.Find(edge.From) != k.uf.Find(edge.To) {
			k.uf.Union(edge.From, edge.To)
			mst = append(mst, edge)
		}
	}
	return mst
}
func NewKruskal[V number, W number](vertices []V) *Kruskal[V, W] {
	uf := NewUnionFind[V](vertices)
	return &Kruskal[V, W]{
		edges: []Edge[V, W]{},
		uf:    uf,
	}
}

type Prim[V number, W number] struct {
	adjacency map[V]map[V]W // adjacency list with weights
	visited   map[V]bool
	minHeap   *Heap[Edge[V, W]]
}
func (p *Prim[V, W]) AddEdge(from, to V, weight W) {
	if p.adjacency == nil {
		p.adjacency = make(map[V]map[V]W)
	}
	if p.adjacency[from] == nil {
		p.adjacency[from] = make(map[V]W)
	}
	if p.adjacency[to] == nil {
		p.adjacency[to] = make(map[V]W)
	}
	p.adjacency[from][to] = weight
	p.adjacency[to][from] = weight
}
func (p *Prim[V, W]) GetMST() []Edge[V, W] {
	if len(p.adjacency) == 0 {
		return nil
	}

	p.visited = make(map[V]bool)

	var startVertex V
	for v := range p.adjacency {
		startVertex = v
		break
	}

	p.visited[startVertex] = true
	for to, weight := range p.adjacency[startVertex] {
		p.minHeap.PushVal(Edge[V, W]{From: startVertex, To: to, Weight: weight})
	}

	mst := []Edge[V, W]{}
	for p.minHeap.Len() > 0 {
		edge := p.minHeap.PopVal()
		if p.visited[edge.To] {
			continue
		}
		mst = append(mst, edge)
		p.visited[edge.To] = true

		for to, weight := range p.adjacency[edge.To] {
			if !p.visited[to] {
				p.minHeap.PushVal(Edge[V, W]{From: edge.To, To: to, Weight: weight})
			}
		}
	}
	return mst
}
func NewPrim[V number, W number]() *Prim[V, W] {
	mh := &Heap[Edge[V, W]]{
		less: func(a, b Edge[V, W]) bool {
			return a.Weight < b.Weight
		},
	}

	return &Prim[V, W]{
		adjacency: make(map[V]map[V]W),
		visited:   make(map[V]bool),
		minHeap:   mh,
	}
}

// NewMST creates a new MST instance based on v, e counts
func NewMST[V number, W number](v_cnt, e_cnt uint) MST[V, W] {
	v_log_floor := bits.Len(v_cnt) - 1
	if e_cnt > v_cnt*uint(v_log_floor) {
		// dense graph
		return NewPrim[V, W]()
	} else {
		// sparse graph
		vertices := make([]V, 0, v_cnt)
		for i := uint(0); i < v_cnt; i++ {
			var vertex V
			vertices = append(vertices, vertex)
		}
		return NewKruskal[V, W](vertices)
	}
}